daemon/cmd/local_node_init.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	k8sLabels "k8s.io/apimachinery/pkg/labels"
+
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/node/addressing"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type localNodeInitializerParams struct {
+	cell.In
+
+	Config    *option.DaemonConfig
+	Clientset client.Clientset
+	LocalNode k8s.LocalNodeResource
+}
+
+// localNodeInitializer performs the bootstrapping of the LocalNodeStore,
+// which contains information about the local Cilium node populated from
+// configuration and Kubernetes.
+type localNodeInitializer struct {
+	localNodeInitializerParams
+}
+
+func (ini *localNodeInitializer) InitLocalNode(ctx context.Context, n *node.LocalNode) error {
+	n.Source = source.Local
+
+	if err := ini.initFromConfig(ctx, n); err != nil {
+		return err
+	}
+
+	if err := ini.initFromK8s(ctx, n); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newLocalNodeInitializer(p localNodeInitializerParams) node.LocalNodeInitializer {
+	return &localNodeInitializer{p}
+}
+
+func (ini *localNodeInitializer) initFromConfig(ctx context.Context, n *node.LocalNode) error {
+	// If there is one device specified, use it to derive better default
+	// allocation prefixes
+	node.SetDefaultPrefix(ini.Config, ini.Config.DirectRoutingDevice, n)
```

Comment:
DirectRoutingDevice is not detected yet here.

Reply:
Ah, I was going to ask how `configureIPAM()` -> `AutoComplete()` ties into this and I guess that's related to your comment here.

Reply:
The original code also first initialized the prefix before auto-detection and then it fixed it up later. So no change to earlier.